src/main/java/io/github/eliux/mega/LocalDateRange.java

```diff
@@ -0,0 +1,45 @@
+package io.github.eliux.mega;
+
+import java.time.LocalDate;
+import java.util.Optional;
+
+/**
+ *This factory is meant for creating DateRange
+ */
+public class LocalDateRange {
```

Comment:
I actually use this class to force the user in using the right format for the date in MegaCmdExport.java

```
  public MegaCmdExport setExpireDate(DateRange dateRange) {
    this.expireLocalDateRange = Optional.of(dateRange);
    return this;
  }
```
if i remove this class, i will have to make this function take a string as parameter. by doing that, i can't ensure that the user actually use the right string format which is "%dy%dm%dd".

@EliuX how do you thing is a better way of doing this?

Reply:
Sure, the class `MegaCmdExport` is important. I recommended not to use a class you had before and you already changed it. Maybe after you applied some changes the comment you read made no sense.
Regarding to this function:
You should name the setter consistently with the name of the internal variable. So instead of `expireLocalDateRange` it should be `expireDate`. Dont worry about describing that it is a `LocalDate`. The fact that this is using `LocalDate` instead of `Date` is because the latest is just the legacy way, but basically the goal is the same. 
Also `Optional#of` is only used when you know that the specified value is not nullable, like `Optional.of(Math.PI).map(...)`. Otherwise you should always use `Optional#ofNullable`.

Reply:
Regarding to this class:
Firstly, the name of this class is missleading. It looks like `DateRange` is storing the `startDate` and the `endDate` with `Date` and LocalDateRange is doing the same but with `LocalDate`, which is not the case because this class is more like a Factory.
Secondly, the problem you are trying to solve is
>  i will have to make this function take a string as parameter. by doing that, i can't ensure that the user actually use the right string format which is "%dy%dm%dd".

If you check it out a little bit, `%dy%dm%dd` is actually the date not a date range. The class `DateRange` should only be in charge of having an `startDate` and an `endDate` and maybe functions to create DateRange instances easily as  [explained before](https://github.com/EliuX/MEGAcmd4J/pull/27/files/972fff2f63b455a2c8683ad230aabfb79bea9857#r471150083). The perfect place/class for creating dates in this particular case would be `java.time.LocalDate`.
Such class is something you cannot change, but those who created it agreed that it should work as its own factory. That is why it has the function:

```
java.time.LocalDate#parse(java.lang.CharSequence, java.time.format.DateTimeFormatter)
```

so, as you are worried about that the format is the one that the import command provides, then you just need to specify the right `DateTimeFormatter` instance.

```
final LocalDate startDate = LocalDate.parse(startDateStr, DateTimeFormatter.BASIC_ISO_DATE)
final LocalDate endDate = LocalDate.parse(endDateStr, DateTimeFormatter.BASIC_ISO_DATE)
```
As you see this [DateTimeFormatter](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#BASIC_ISO_DATE) already has a predefined preset for [calendar dates in format ISO 8601](https://en.wikipedia.org/wiki/ISO_8601#Calendar_dates)
and it is easier that if MegaCMD changes the format we just have to change the instance of the `DateTimeFormatter` parameter.

This is why that besides having a bad name and responsibility, this class should not exist.

Reply:
i get you but the problem is that `%dy%dm%dd` is not actually the date.
it's a delay, the number of years,months,days that the link will stay available.
here is what the official documentation says 

`--expire=TIMEDELAY     Determines the expiration time of a node.
                           It indicates the delay in hours(h), days(d),
                           minutes(M), seconds(s), months(m) or years(y)
                           e.g. "1m12d3h" establish an expiration time 1 month,
                           12 days and 3 hours after the current moment`

One simple solution would be to create a function that will have the delays directly as a parameter, it will be simple.
But i wanted to make something more user friendly by providing a builder that will calculate that range (delay). That's why i used this classes, i understand the name is misleading we can talk about `time delay` istead of range.

we can do something like this in `MegaCmdExport.java`:

```
  public MegaCmdExport setExpireDate(String dateDelay) {
    validate(dateDelay)
    this.expireDate = Optional.ofNullable(dateRange);
    return this;
  }
```

validate() will make sure the format is `%dy%dm%dd` otherwise it will throw an exception
then when we execute the cmd we just add directly the string 

```
 expireDate .ifPresent(date -> cmdParams.add("--expire=" + date));
```
in this case we could remove those two classes DateRange and LocalDateRange
Eventhough this is simpler the user would not know which excat format we are waiting for.

that's why i introduce LocalDateRange and DateRange for the application to calculate the String of the delay, based on the startDate en EndDate.

LocalDateRange (i can rename it DateRangeBuilder) is just a builder who will build a DateRange based on a startDate and endDate.
The method `toString()` of the DateRange generates a string with the right format `%dy%dm%dd`.
so in `MegaCmdExport.java`, i was able to do this

```
 expireLocalDateRange.ifPresent(dateRange -> cmdParams.add("--expire=" + dateRange.toString()));
```

Reply:
If this is a delay, then why the the class name is `LocalDateRange`. IMHO, the best way is that you can add any time of delay to a Date and find a type/class which can represent this offset. Otherwise the function `setExpireDate` will be ambiguous: you may expect that you would be adding a date in String format there. Also, that function would have double responsibility, to convert that `String` to a `Date`, which is basically parsing and assign that the result `Date` to the attribute `expireDate`.  The best type/class that could store such time delay would be Duration, i.e.

```  
    TemporalAmount delay = Duration.of(1, ChronoUnit.DAYS);
    LocalDateTime endDate =  LocalDateTime.now();
    endDate.plus(delay);
```
I am not against `DateRange`, but `LocalDateRange` which seems missleading.
Regarding to

> LocalDateRange (i can rename it DateRangeBuilder) is just a builder who will build a DateRange based on a startDate and endDate.

Why DateRange which is a custom class cannot be its own Factory  as `Optional`, `LocalDate` and many others? Adding a Builder would make sense if the building process would be really complex, then something like a [StepBuilder ](https://softwareengineering.stackexchange.com/questions/324316/java-how-to-implement-a-step-builder-for-which-the-order-of-setters-doesnt-mat)would make sense.

Regarding to:

> The method toString() of the DateRange generates a string with the right format %dy%dm%dd.

This would assume that `DateRange` would only be represented to `String` this way, because the logic you are using it right now requires it so. What if it just stores the data and another function provides that representation? That would make it flexible for any case right?

Another thing I see confusing is that the time delay for expiration is a `DateRange`. Such delay as you mentioned is

> the delay in hours(h), days(d), minutes(M), seconds(s), months(m) or years(y)

A DateRange should indicate an start date and an end Date. For instance you can create and `DateRange` using a start `LocalDate` and a `Duration`  by adding that start date plus that duration, right?

So a line like 

>  expireLocalDateRange.ifPresent(dateRange -> cmdParams.add("--expire=" + dateRange.toString()));

seems really confusing because I would be expecting that the after `--expire` I see something like `1m12d3h` which seems like a representation of a `Duration`, right? I even realised that there was also a `Period` class which basically store less data, but the necessary we need. As it says in its documentation:

> This class models a quantity or amount of time in terms of years, months and days.
See `Duration` for the time-based equivalent to this class.

So, a DateRange could have a convenient `getPeriod` function which could return a `period` attribute or calculate it based on the end and a start `LocalDate`:

```
public Duration getDuration() {
  return period.orElseGet(()-> startDate.until(endDate));
}
```

But what this logics tells me is that you not even really have a `DateRange`, because the endDate is uncertain, because if the expiration date or the duration is not set then you have just a start date. Then I had to see the help for this command to see why it was getting so complicated and I saw that they even had a way for the user to specify the format of the date. So this should be simpler.

Reply:
Let me propose you one thing. Let me create the `MegaCmdExport` class and resolve the type of data and converters this would have to keep the code simple and clean. I will merge it to `develop` and I will let you create the tests to see wheter if it passes the crtiteria you need. This time you would review what I did and let me know if I am missing any use case. This will be a team work ;). So, please stop the development right now and I will start working on this ASAP. I will tag you for a revision and then you apply the tests and any changes you see fit.